errors/type.go

```
@@ -0,0 +1,161 @@
+	case Internal:
+		return "Internal"
+	case InvalidArgument:
+		return "Invalid argument"
+	case OutOfRange:
+		return "Out of range"
+	case NotFound:
+		return "Not found"
+	case Conflict:
+		return "Conflict"
+	case AlreadyExists:
+		return "Already exists"
+	case Unauthorized:
+		return "Unauthorized"
+	case PermissionDenied:
+		return "Permission denied"
+	case Timeout:
+		return "Timeout"
+	case NotImplemented:
+		return "Not implemented"
+	case TemporarilyUnavailable:
+		return "Temporarily unavailable"
+	case PermanentlyUnavailable:
+		return "Permanently unavailable"
+	case Canceled:
+		return "Canceled"
+	case ResourceExhausted:
+		return "Resource exhausted"
+	default:
+		return "Unknown"
+	}
+}
+
+// MarshalText implements TextMarsheler
+func (t Type) MarshalText() ([]byte, error) {
+	return []byte(t.String()), nil
+}
+
+// UnmarshalText implements TextUnmarsheler
+func (t *Type) UnmarshalText(text []byte) error {
+	e, err := fromString(string(text))
+	if err != nil {
+		return err
+	}
+
+	*t = e
+
+	return nil
+}
+
+func fromString(str string) (Type, error) {
+	enum := strings.ToLower(str)
+	switch enum {
+	case "unknown":
+		return Unknown, nil
+	case "internal":
+		return Internal, nil
+	case "invalid argument":
+		return InvalidArgument, nil
+	case "out of range":
+		return OutOfRange, nil
+	case "not found":
+		return NotFound, nil
+	case "conflict":
+		return Conflict, nil
+	case "already exists":
+		return AlreadyExists, nil
+	case "unauthorized":
+		return Unauthorized, nil
+	case "permission denied":
+		return PermissionDenied, nil
+	case "timeout":
+		return Timeout, nil
+	case "not implemented":
+		return NotImplemented, nil
+	case "temporarily unavailable":
+		return TemporarilyUnavailable, nil
+	case "permanently unavailable":
+		return PermanentlyUnavailable, nil
+	case "canceled":
+		return Canceled, nil
+	case "resource exhausted":
+		return ResourceExhausted, nil
+	default:
+		return Unknown, fmt.Errorf("Invalid event type")
```

not an event, also, indicate somewhere that the error can be ignored